pkg/cache/application.go

```diff
@@ -58,6 +58,7 @@ type Application struct {
 	placeholderAsk             *si.Resource // total placeholder request for the app (all task groups)
 	placeholderTimeoutInSec    int64
 	schedulingStyle            string
+	requestOriginatingPod      *v1.Pod // Original Pod which creates the requests
```

Comment:
Since we need Pod, tried to set and use the Pod only. But, I can see your point - relaying on *Task is much better as it can be used for other functions as well. Taken care.

Reply:
This is strange. I understand the intention was to add some pod events to expose that placeholder times out; however, sending this event whenever we delete a placeholder to the first pod of the app is strange...
if we have an app CRD, that might be a better place to gather the event; but today we do not have that. I suggest to just print some logs for now. unless there is a better place to expose such messages.

Reply:
I do not think you have read the code correctly. See the code in `publishPlaceholderTimeoutEvents()`

The event is only written to the first pod if the pod that is finished is a placeholder and the TerminationType is a timeout. That means it will only happen for placeholders that were requested but not replaced. If and when you get the app CRD for all that can be easily changed to be set on the CRD.
For gang scheduling the first pod contains the annotations. It will always be set as the originating pod for the app. Placeholder timeouts will thus be linked to the pod with the annotations.
I don't think we have made this a generic any finished pod to a random pod kind of setup.

Reply:
If we submit a spark job with GS enabled, 1 driver + 5 executors like this:

```
spark-1234-driver
tg-spark-1234-executor-0
tg-spark-1234-executor-1
tg-spark-1234-executor-2
tg-spark-1234-executor-3
tg-spark-1234-executor-4
```
if the placeholders timeout we push all events to the driver pod
```
spark-1234-driver
Events
  Application spark-1234 placeholder has been timed out, task: taskIDString0
  Application spark-1234 placeholder has been timed out, task: taskIDString1
  Application spark-1234 placeholder has been timed out, task: taskIDString2
  ...
```
if we have 100 executors timeout, we will have 100 events like this printed in the driver pod, that won't help the users..
when we have spark-operator running, it makes more sense to aggregate such messages and send them to the spark CRD object: https://github.com/apache/incubator-yunikorn-k8shim/blob/master/pkg/appmgmt/sparkoperator/spark.go , sending to the driver pod with such messages does not seem to be very helpful to me.

Reply:
I would say it does help. Even if they get annoyed by the fact that there are a lot of events it tells them that they need to look at their gang request. Something is wrong if there are lots of time out messages. Either with their request or the app or the cluster.

If we have the possibility to update the CRD then we can do that. If we do not have that possibility now then that is not an option. We also need to handle the case that we have a gang scheduling request that does not use the operator. In that case the only reasonable way to tell the user that the placeholders are timing out is via the driver pod.

Reply:
@wilfred-s @yangwwei Thanks for sharing your thoughts.

Given the current limitation in some cases and future plans, Can we do the following?

1. Filing a separate JIRA for aggregation in general for all places (not just only for this particular case). This JIRA can focus mainly on how to make life easier on monitoring aspects by aggregating all such events. Also, destination has to be fixed across different scenarios because expecting Admin to go over 'n' number of different places to look for events is difficult for them during troubleshooting.
2. Filing a separate JIRA to take care of these kind of events notifications wherever we have operator integration as of now. If needed we can do this. Otherwise, we can cover this in above JIRA itself.
3. Given the above plans, Can we push this PR as an interim solution?

Reply:
Sorry, I do not agree to add this.
My concern is such events may not help and even worse they may confuse users. 
I prefer not to add this now until we have a better solution.

Reply:
Ok. I will think through further on doing this in a better way.

Reply:
I think we can make this less confusing for users if we ensure that all the placeholder timeout events are the same (i.e. don't include the placeholder ID). This way, the kubernetes UI tools will coalesce those (100) events down to a single entry with a counter. This is very useful and not at all overwhelming.

Reply:
Taken care